app/Console/Commands/CleanupDatabase.php

```diff
@@ -0,0 +1,80 @@
+        self::delete_unused_rows('user2project', 'projectid', 'project');
+        self::delete_unused_rows('userstatistics', 'projectid', 'project');
+
+        self::delete_unused_rows('build2configure', 'buildid', 'build');
+        self::delete_unused_rows('build2note', 'buildid', 'build');
+        self::delete_unused_rows('build2test', 'buildid', 'build');
+        self::delete_unused_rows('buildemail', 'buildid', 'build');
+        self::delete_unused_rows('builderror', 'buildid', 'build');
+        self::delete_unused_rows('builderrordiff', 'buildid', 'build');
+        self::delete_unused_rows('buildfailure', 'buildid', 'build');
+        self::delete_unused_rows('buildfailuredetails', 'id', 'buildfailure', 'detailsid');
+        self::delete_unused_rows('buildinformation', 'buildid', 'build');
+        self::delete_unused_rows('buildtesttime', 'buildid', 'build');
+        self::delete_unused_rows('configure', 'id', 'build2configure', 'configureid');
+        self::delete_unused_rows('configureerror', 'configureid', 'configure');
+        self::delete_unused_rows('configureerrordiff', 'buildid', 'build');
+        self::delete_unused_rows('coverage', 'buildid', 'build');
+        self::delete_unused_rows('coveragefilelog', 'buildid', 'build');
+        self::delete_unused_rows('coveragesummary', 'buildid', 'build');
+        self::delete_unused_rows('coveragesummarydiff', 'buildid', 'build');
+        self::delete_unused_rows('dailyupdatefile', 'dailyupdateid', 'dailyupdate');
+        self::delete_unused_rows('dynamicanalysis', 'buildid', 'build');
+        self::delete_unused_rows('label2build', 'buildid', 'build');
+        self::delete_unused_rows('note', 'id', 'build2note', 'noteid');
+        self::delete_unused_rows('subproject2build', 'buildid', 'build');
+        self::delete_unused_rows('summaryemail', 'buildid', 'build');
+        self::delete_unused_rows('testdiff', 'buildid', 'build');
+        self::delete_unused_rows('testoutput', 'id', 'build2test', 'outputid');
+        self::delete_unused_rows('updatefile', 'updateid', 'buildupdate');
+        self::delete_unused_rows('uploadfile', 'id', 'build2uploadfile', 'fileid');
+
+        self::delete_unused_rows('dynamicanalysisdefect', 'dynamicanalysisid', 'dynamicanalysis');
+        self::delete_unused_rows('subproject2subproject', 'subprojectid', 'subproject');
+
+        self::delete_unused_rows('coveragefile', 'id', 'coverage', 'fileid');
+        self::delete_unused_rows('coveragefile2user', 'fileid', 'coveragefile');
+
+        self::delete_unused_rows('test2image', 'outputid', 'testoutput');
+        self::delete_unused_rows('image', 'id', 'test2image', 'imgid');
+    }
+
+    /** Delete unused rows */
+    private static function delete_unused_rows(string $table, string $field, string $targettable, string $selectfield = 'id'): void
+    {
+        DB::delete("DELETE FROM $table WHERE $field NOT IN (SELECT $selectfield AS $field FROM $targettable)");
```

Comment:
Should this delete in batches too?  You can add a `LIMIT` clause to the end, and then iterate until no more results are deleted (`DB::delete` returns the number of rows deleted).

Reply:
It turns out postgres [does not support](https://github.com/Kitware/CDash/actions/runs/11112806418/job/30875687907?pr=2468#step:5:3155) `DELETE FROM ... LIMIT`, so we would need to do something more clever here to allow for deleting in batches.

Reply:
You could try `DELETE FROM ... WHERE ... NOT IN (SELECT ... LIMIT ...)` instead.  I'm not sure whether the optimizer would be smart enough to do it optimally, but it's probably worth a shot.

Reply:
Wouldn't that result in data loss? The inner `SELECT` is telling us what rows to keep. If we put a `LIMIT 1` (eg.) there, the outer DELETE would then remove all but one record from the table, right?

Reply:
Oops I mistyped the pseudo-query.  See edit.